README.md
# node-red-contrib-kubernetes-client

A `Node-RED` node that supports interacting with Kubernetes API via `watches` and `HTTP` requests.

[GitHub](https://github.com/travisghansen/node-red-contrib-kubernetes-client)

# Install

```
npm install --save node-red-contrib-kubernetes-client
```

# Documentation

The nodes are properly documented in `Node-RED` itself. In short there are 2 nodes:

- `kubernetes-client-watch` - produces messages for configured `watch` endpoints.
- `kubernetes-client-http` - allows complete interaction with Kubernetes API via **ALL** HTTP `endpoints` and `verbs` etc.

`Watches` ouput a `msg.payload` with the following structure (as a `json` object in `Node-RED` but shared here as `yaml` for readability):

```
# example event
---
type: ADDED || MODIFIED || DELETED || ERROR
object:
  kind: ...
  apiVersion: ...
  metadata:
    ...
  ...

# example error
---
type: ERROR
object:
  kind: Status
  apiVersion: v1
  metadata: {}
  status: Failure
  message: 'too old resource version: 1 (78390381)'
  reason: Gone
  code: 410

```

`msg.payload.object` contains the full resource from Kubernetes.

# Development

Some helpful command variants for testing/developing:

```
# show event output structure
kubectl -v6 get nodes --watch --output-watch-events -o yaml

# manually specify URL
kubectl -v6 get --raw '/api/v1/nodes?resourceVersion=1&watch=true'

# watch response error structure
---
type: ERROR
object:
  kind: Status
  apiVersion: v1
  metadata: {}
  status: Failure
  message: 'too old resource version: 1 (78383979)'
  reason: Gone
  code: 410

```

# License

See [license](https://github.com/travisghansen/node-red-contrib-kubernetes-client/blob/master/LICENSE) (MIT).


CHANGELOG.md
# v0.2.0

Support easier way to set/retrieve object endpoints.

- `topic` for the `http` node can be a full resource object
- attempt to add `selfLink` property to `Event`s `involvedObject`.

# v0.1.0

Initial release:

- support watches
- support http endpoints


src/config.js
"use strict";
const k8s = require("@kubernetes/client-node");
const LRU = require("lru-cache");
const queryString = require("query-string");
const request = require("request");
const URI = require("uri-js");

const FAILURE_CACHE_TIME = 5 * 60 * 1000; // 5 minutes
const SUCCESS_CACHE_TIME = 1 * 60 * 60 * 1000; // 1 hour
class KubeConfig extends k8s.KubeConfig {
  constructor() {
    super(...arguments);
    this.discoveryCache = new LRU({ maxAge: SUCCESS_CACHE_TIME });
  }

  /**
   * creates a new watch instance
   */
  createWatch() {
    return new k8s.Watch(this);
  }

  /**
   * msg.method
   * msg.topic
   * msg.payload
   *
   * @param {*} msg
   */
  async makeHttpRestRequest(msg) {
    const kc = this;
    return new Promise(async (resolve, reject) => {
      let endpoint = msg.topic;
      if (typeof endpoint == "object") {
        try {
          endpoint = await this.buildResourceSelfLink(msg.topic);
        } catch (err) {
          reject(err);
        }
      }
      endpoint = kc.buildWatchlessURI(endpoint);
      msg.method = msg.method || "GET";
      const options = {
        method: msg.method.toUpperCase(),
        url: `${kc.getCurrentCluster().server}${endpoint}`,
        headers: {
          Accept: "application/json",
          "User-Agent": "Node-RED",
          "Content-Type": "application/json"
        },
        json: true
        //agentOptions: {
        //  rejectUnauthorized: false
        //}
      };

      kc.applyToRequest(options);

      if (msg.method.includes("PATCH")) {
        /**
         * https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#patch-operations
         * https://github.com/kubernetes/community/blob/master/contributors/devel/strategic-merge-patch.md
         *
         * Content-Type: application/json-patch+json
         * Content-Type: application/merge-patch+json
         * Content-Type: application/strategic-merge-patch+json
         */
        switch (msg.method) {
          case "PATCH-JSON":
            options["headers"]["Content-Type"] = "application/json-patch+json";
            break;
          case "PATCH-STRATEGIC-MERGE":
            options["headers"]["Content-Type"] =
              "application/strategic-merge-patch+json";
            break;
          case "PATCH":
          case "PATCH-MERGE":
          default:
            options["headers"]["Content-Type"] = "application/merge-patch+json";
            break;
        }

        options.method = "PATCH";
      }

      switch (options.method.toUpperCase()) {
        case "GET":
          options.qs = msg.payload;
          break;
        default:
          options.body = msg.payload;
          break;
      }

      request(options, function(err, res, body) {
        if (err) {
          reject(err);
        }
        resolve(res);
      });
    });
  }

  /**
   * Given a URI, remove parts of the path that equal 'watch'
   * and also remove any 'watch' parameters from the query string
   *
   * @param {*} uri
   */
  buildWatchlessURI(uri) {
    const suri = URI.parse(URI.normalize(uri));
    const path = suri.path;
    const query = suri.query;

    const pathParts = path.split("/");
    const pathPartsWithoutWatch = pathParts.filter(item => {
      if (item.toLowerCase() != "watch") {
        return true;
      }
      return false;
    });

    const squery = queryString.parse(query);
    delete squery["watch"];

    const newPath = pathPartsWithoutWatch.join("/");
    const newUri = URI.serialize({
      path: newPath,
      query: queryString.stringify(squery)
    });

    return newUri;
  }

  async getAPIGroups() {
    let res;
    let cacheKey = "__APIGroups";
    res = this.discoveryCache.get(cacheKey);
    if (res === undefined) {
      res = await this.makeHttpRestRequest({
        topic: "/apis"
      });

      if (res.stausCode == 200) {
        res = res.body;
        this.discoveryCache.set(cacheKey, res);
      } else {
        res = res.body;
        this.discoveryCache.set(cacheKey, res, FAILURE_CACHE_TIME);
      }
    }

    return res;
  }

  async getAPIResources(preferredVersions = false) {
    this.locks = this.locks || {};
    let cacheKey = "__APIResources";

    function sleep(ms) {
      return new Promise(resolve => setTimeout(resolve, ms));
    }

    let i = 0;
    while (this.locks[cacheKey]) {
      i++;
      await sleep(1000);

      if (i > 10) {
        break;
      }
    }

    // test cache lock and wait
    let resources = this.discoveryCache.get(cacheKey);

    if (resources === undefined) {
      // apply lock
      this.locks[cacheKey] = true;

      resources = [];

      try {
        let apiGroups = await this.getAPIGroups();
        let api = await this.makeHttpRestRequest({ topic: "/api" });

        if (api.statusCode == 200) {
          await Promise.all(
            api.body.versions.map(async version => {
              let res = await this.makeHttpRestRequest({
                topic: `/api/${version}`
              });

              if (res.statusCode == 200) {
                resources.push(res.body);
              }
            })
          );
        }

        if (apiGroups.groups) {
          await Promise.all(
            apiGroups.groups.map(async group => {
              await Promise.all(
                group.versions.map(async version => {
                  let res = await this.makeHttpRestRequest({
                    topic: `/apis/${version.groupVersion}`
                  });

                  if (res.statusCode == 200) {
                    resources.push(res.body);
                  }
                })
              );
            })
          );
        }
      } catch (err) {}

      if (resources.length > 0) {
        this.discoveryCache.set(cacheKey, resources);
      }

      // remove lock
      this.locks[cacheKey] = false;
    }

    if (preferredVersions) {
      let apiGroups = await this.getAPIGroups();
      resources = resources.filter(resource => {
        // apiVersion is not present on core api resource list
        if (!resource.apiVersion) {
          return true;
        }
        return apiGroups.groups.some(group => {
          return group.preferredVersion.groupVersion == resource.groupVersion;
        });
      });
    }

    return resources;
  }

  /**
   * 
   * @param {*} kind 
   * @param {*} version 
   */
  async getApiGroupVersion(kind, version) {
    const resources = await this.getAPIResources(true);
    let matches = resources.filter(resourceList => {
      let groupVersionVersion = resourceList.groupVersion.split("/").pop();
      if (version && !(groupVersionVersion == version)) {
        return false;
      }

      return resourceList.resources.some(resource => {
        return resource.kind.toLowerCase() == kind.toLowerCase();
      });
    });

    if (matches.length == 1) {
      return matches[0].groupVersion;
    }
  }

  async buildResourceSelfLink(kind, apiVersion, name, namespace) {
    let res;
    if (arguments.length == 1) {
      res = kind;

      if (res.metadata && res.metadata.selfLink) {
        return res.metadata.selfLink;
      }

      kind = res.kind;
      apiVersion = res.apiVersion;
      if (res.metadata) {
        name = res.metadata.name || res.name;
        namespace = res.metadata.namespace || res.namespace || null;
      } else {
        // this support the involvedObject syntax of Events
        name = res.name;
        namespace = res.namespace || null;
      }
    }

    // for testing
    //apiVersion = undefined;

    if (!apiVersion) {
      apiVersion = await this.getApiGroupVersion(kind);
    }

    if (apiVersion == undefined || apiVersion === null) {
      throw new Error("missing apiVersion");
    }

    let prefix = "/apis";
    if (apiVersion == "v1") {
      prefix = "/api";
    }

    //console.log(kind, apiVersion, name, namespace);

    /**
     * cache each unique resource as a key
     * allows for JIT lookup and freshness on a per-resource basis
     */
    let cacheKey = `${prefix}/${apiVersion}`;
    res = this.discoveryCache.get(cacheKey);
    if (res === undefined) {
      res = await this.makeHttpRestRequest({
        topic: `${prefix}/${apiVersion}`
      });

      if (res.statusCode == 200) {
        res = res.body;
        this.discoveryCache.set(cacheKey, res);
      } else if (res.statusCode == 404) {
        // assume in incomplete apiVersion (ie: only the version and not groupVersion)
        // attempt to find full groupVersion

        let newApiVersion = await this.getApiGroupVersion(kind, apiVersion);
        // try again
        if (newApiVersion && newApiVersion != apiVersion) {
          apiVersion = newApiVersion;
          res = await this.makeHttpRestRequest({
            topic: `${prefix}/${apiVersion}`
          });

          if (res.statusCode == 200) {
            res = res.body;
            this.discoveryCache.set(cacheKey, res);
          } else {
            res = res.body;
            this.discoveryCache.set(cacheKey, res, FAILURE_CACHE_TIME);
          }
        } else {
          res = res.body;
          this.discoveryCache.set(cacheKey, res, FAILURE_CACHE_TIME);
        }
      } else {
        res = res.body;
        this.discoveryCache.set(cacheKey, res, FAILURE_CACHE_TIME);
      }
    }

    if (!res.resources) {
      return;
    }

    let resource = res.resources.find(resource => {
      return resource.kind == kind;
    });

    if (resource) {
      let endpoint = "";
      if (resource.namespaced) {
        endpoint = `${prefix}/${apiVersion}/namespaces/${namespace}/${resource.name}/${name}`;
      } else {
        endpoint = `${prefix}/${apiVersion}/${resource.name}/${name}`;
      }

      return endpoint;
    } else {
      throw new Error("failure to lookup resource selfLink");
    }
  }

  async dressEventResource(event, removeOld = true) {
    if (!event) {
      return;
    }
    if (!event.involvedObject) {
      return;
    }

    // move traditional metadata fields to metadata block
    event.involvedObject.metadata = event.involvedObject.metadata || {};
    [
      "name",
      "namespace",
      "uid",
      "resourceVersion",
      "creationTimestamp"
    ].forEach(attribute => {
      if (
        event.involvedObject.hasOwnProperty(attribute) &&
        !event.involvedObject.metadata.hasOwnProperty(attribute)
      ) {
        event.involvedObject.metadata[attribute] =
          event.involvedObject[attribute];

        if (removeOld) {
          delete event.involvedObject[attribute];
        }
      }
    });

    // attempt to cleanup incomplete apiVersion (ie: not including the group)
    // if applicable
    let newApiVersion = await this.getApiGroupVersion(
      event.involvedObject.kind,
      event.involvedObject.apiVersion
    );

    if (newApiVersion) {
      if (!event.involvedObject.apiVersion) {
        event.involvedObject.apiVersion = newApiVersion;
      } else if (event.involvedObject.apiVersion != newApiVersion) {
        event.involvedObject.apiVersion = newApiVersion;
      }
    }

    // attempt to create selfLink
    if (!event.involvedObject.metadata.hasOwnProperty("selfLink")) {
      try {
        let selfLink = await this.buildResourceSelfLink(event.involvedObject);

        event.involvedObject.metadata.selfLink = selfLink;
      } catch (err) {}
    }
  }
}

module.exports.KubeConfig = KubeConfig;


src/index.js
/**
 * https://github.com/kubernetes-client/javascript/issues/392
 *
 */
module.exports = function(RED) {
  "use strict";
  const KubeConfig = require("./config").KubeConfig;

  /**
   * https://nodered.org/docs/creating-nodes/status
   * The shape property can be: ring or dot
   * The fill property can be: red, green, yellow, blue or grey
   */
  const statuses = {
    connecting: { fill: "yellow", shape: "ring", text: "connecting" },
    connected: { fill: "green", shape: "dot", text: "connected" },
    disconnected: { fill: "red", shape: "ring", text: "disconnected" },
    misconfigured: { fill: "red", shape: "ring", text: "misconfigured" },
    error: { fill: "red", shape: "dot", text: "error" },
    sending: { fill: "blue", shape: "dot", text: "sending" },
    receiving: { fill: "blue", shape: "dot", text: "receiving" },
    transfer: { fill: "blue", shape: "dot", text: "transfer" },
    blank: {}
  };

  function KubernetesClientConfigNode(n) {
    RED.nodes.createNode(this, n);
    this.options = {};
    this.kc = new KubeConfig();

    if (this.credentials.kubeConfig) {
      this.kc.loadFromString(this.credentials.kubeConfig);
    } else {
      this.kc.loadFromDefault();
    }

    this.on("close", (removed, done) => {
      if (removed) {
        // This node has been deleted
      } else {
        // This node is being restarted
      }
      this.kc.discoveryCache.reset();
      delete this.kc.discoveryCache;
      delete this.kc;

      done();
    });
  }

  RED.nodes.registerType(
    "kubernetes-client-config",
    KubernetesClientConfigNode,
    {
      credentials: {
        kubeConfig: { type: "text" }
      }
    }
  );

  /**
   * TODO: could have more intelligent reconnect strategy
   *
   * @param {*} n
   */
  function KubernetesClientWatchNode(n) {
    RED.nodes.createNode(this, n);

    this.options = {
      endpoint: n.endpoint,
      activityTimeout:
        parseInt(n.activityTimeout) >= 0 ? parseInt(n.activityTimeout) : 90,
      initialResourceVersionStrategy: n.initialResourceVersionStrategy || "",
      goneResourceVersionStrategy: n.goneResourceVersionStrategy || ""
    };

    const node = this;
    node.lastMessageTimestamp = 0;
    this.kubernetesClientConfig = n.kubernetesClientConfig;
    this.kubernetesClientConfigNode = RED.nodes.getNode(
      this.kubernetesClientConfig
    );

    const kc = this.kubernetesClientConfigNode.kc;
    const watch = kc.createWatch();
    const endpoint = node.options.endpoint;
    const endpointHash = require("crypto")
      .createHash("md5")
      .update(`${node.kubernetesClientConfig}:${endpoint}`)
      .digest("hex");
    let endpointHashHasBeenSet = false;
    let connecting = false;
    let resourceVersion = false;
    let latestResourceVersion = null;
    let forcedResourceVersion = false;
    let triggerWatchViaShortInterval = false;
    let shortIntervalSeconds = 10;

    node.startWatch = async function() {
      if (node.watch) {
        node.watch.abort();
        node.watch.destroy();
        delete node.watch;
      }

      if (connecting == true) {
        return;
      }

      node.status(statuses.connecting);
      connecting = true;

      /**
       * https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions
       * 0 vs null have special meaning with watches, carefully choose
       */
      if (forcedResourceVersion !== false) {
        resourceVersion = forcedResourceVersion;
        forcedResourceVersion = false;
      } else if (latestResourceVersion) {
        resourceVersion = latestResourceVersion;
      } else if (resourceVersion !== false) {
        resourceVersion = resourceVersion;
      } else {
        switch (node.options.initialResourceVersionStrategy) {
          case "CURRENT":
            try {
              const res = await kc.makeHttpRestRequest({
                topic: endpoint,
                payload: { limit: 1 }
              });

              if (res.statusCode == 200) {
                resourceVersion = res.body.metadata.resourceVersion;
              } else {
                resourceVersion = null;
              }
            } catch (err) {
              node.error(err);
              resourceVersion = null;
            }

            break;
          case "NULL":
            resourceVersion = null;
            break;
          case "ZERO":
            resourceVersion = 0;
            break;
          case "RESTORE-NULL":
            if (endpointHash == node.context().get("endpointHash")) {
              resourceVersion = node.context().get("resourceVersion") || null;
            } else {
              resourceVersion = null;
            }
            break;
          case "RESTORE-ZERO":
            if (endpointHash == node.context().get("endpointHash")) {
              resourceVersion = node.context().get("resourceVersion") || null;
            } else {
              resourceVersion = 0;
            }
            break;
          case "RESTORE-CURRENT":
          default:
            if (endpointHash == node.context().get("endpointHash")) {
              resourceVersion = node.context().get("resourceVersion") || null;
            } else {
              try {
                const res = await kc.makeHttpRestRequest({
                  topic: endpoint,
                  payload: { limit: 1 }
                });

                if (res.statusCode == 200) {
                  resourceVersion = res.body.metadata.resourceVersion;
                } else {
                  resourceVersion = null;
                }
              } catch (err) {
                node.error(err);
                resourceVersion = null;
              }
            }
            break;
        }
      }

      // final sanity check
      if (resourceVersion !== null && !(resourceVersion >= 0)) {
        resourceVersion = null;
      }

      node.log(
        `watching ${
          kc.getCurrentCluster().server
        }${endpoint} from resourceVersion: ${resourceVersion}`
      );

      node.watch = watch.watch(
        endpoint,
        { resourceVersion },
        async (type, object) => {
          if (type === undefined || object === undefined) {
            return;
          }

          node.lastMessageTimestamp = Math.floor(Date.now() / 1000);

          if (type == "ERROR") {
            node.error(
              `kubernetes watch (${
                kc.getCurrentCluster().server
              }${endpoint}) error - status: ${object.status}, message: ${
                object.message
              }, reason: ${object.reason}, code: ${object.code}`
            );
            const status = JSON.parse(JSON.stringify(statuses.error));
            status.text = status.text + ": " + object.message;
            node.status(status);

            if (
              object.code == 410 &&
              ["Gone", "Expired"].includes(object.reason)
            ) {
              forcedResourceVersion = null;
              switch (node.options.goneResourceVersionStrategy) {
                case "ZERO":
                  forcedResourceVersion = 0;
                  break;
                case "NULL":
                  forcedResourceVersion = null;
                  break;
                case "CURRENT":
                default:
                  try {
                    const res = await kc.makeHttpRestRequest({
                      topic: endpoint,
                      payload: { limit: 1 }
                    });

                    if (res.statusCode == 200) {
                      forcedResourceVersion = res.body.metadata.resourceVersion;
                    }
                  } catch (err) {
                    node.error(err);
                  }
                  break;
              }
            }
          }

          if (
            object.metadata.resourceVersion &&
            object.metadata.resourceVersion > Number(latestResourceVersion)
          ) {
            latestResourceVersion = object.metadata.resourceVersion;
            node.context().set("resourceVersion", latestResourceVersion);
            if (endpointHashHasBeenSet === false) {
              node.context().set("endpointHash", endpointHash);
              endpointHashHasBeenSet = true;
            }
          }

          // TODO: perhaps use object.metadata.creationTimestamp to filter out ADDED events on startup

          node.status(statuses.transfer);
          const msg = {};
          msg.payload = { type, object };
          msg.topic = object.metadata.selfLink || "";

          /**
           * try to add selfLink to involvedObject
           */
          if (
            msg.payload.object.kind == "Event" &&
            msg.payload.object.apiVersion == "v1"
          ) {
            try {
              await kc.dressEventResource(msg.payload.object);
            } catch (err) {}
          }

          msg.kube = {};
          msg.kube.config = {};
          msg.kube.config.cluster = kc.getCurrentCluster();
          msg.kube.config.context = kc.getCurrentContext();
          msg.kube.config.user = kc.getCurrentUser();
          msg.kube.client = kc;
          node.send(msg);
          node.status(statuses.connected);
        },
        err => {
          connecting = false;
          node.status(statuses.disconnected);
          triggerWatchViaShortInterval = true;
          if (err) {
            node.error(
              `kubernetes watch (${
                kc.getCurrentCluster().server
              }${endpoint}) error: ${err}`
            );
            const status = JSON.parse(JSON.stringify(statuses.error));
            status.text = status.text + ": " + err;
            node.status(status);

            if (err.code == "ETIMEDOUT") {
              node.log(
                `attempting connect to kubernetes watch (${
                  kc.getCurrentCluster().server
                }${endpoint}) due to connect timeout`
              );
            } else if (err == "resourceVersion stale") {
              node.log(
                `attempting connect to kubernetes watch (${
                  kc.getCurrentCluster().server
                }${endpoint}) due to stale resourceVersion`
              );
            } else {
              node.log(
                `attempting connect to kubernetes watch (${
                  kc.getCurrentCluster().server
                }${endpoint}) due to ${err}`
              );
            }
          } else {
            node.log(
              `attempting connect to kubernetes watch (${
                kc.getCurrentCluster().server
              }${endpoint}) due to unknown connection closure`
            );
          }
        }
      );

      node.watch.on("socket", socket => {});

      node.watch.on("error", err => {});

      node.watch.on("close", () => {});

      node.watch.on("response", response => {
        connecting = false;
        if (response.statusCode == 200) {
          node.status(statuses.connected);
        }
      });
    };

    node.on("close", (removed, done) => {
      if (removed) {
        // This node has been deleted
      } else {
        // This node is being restarted
      }
      if (node.watch) {
        node.watch.abort();
        node.watch.destroy();
        delete node.watch;
      }

      if (node.activityTimeoutInterval) {
        clearInterval(node.activityTimeoutInterval);
      }

      if (node.shortInterval) {
        clearInterval(node.shortInterval);
      }

      done();
    });

    if (node.kubernetesClientConfig) {
      node.startWatch();

      if (node.activityTimeoutInterval) {
        clearInterval(node.activityTimeoutInterval);
      }

      if (node.options.activityTimeout > 0) {
        node.activityTimeoutInterval = setInterval(() => {
          const currentTimestamp = Math.floor(Date.now() / 1000);
          if (
            !connecting &&
            currentTimestamp - node.lastMessageTimestamp >
              node.options.activityTimeout
          ) {
            node.log(
              `attempting reconnect to kubernetes watch (${
                kc.getCurrentCluster().server
              }${endpoint}) due to inactivity timeout`
            );
            node.startWatch();
          }
        }, node.options.activityTimeout * 1000);
      }

      if (node.shortInterval) {
        clearInterval(node.shortInterval);
      }
      node.shortInterval = setInterval(() => {
        if (!connecting && triggerWatchViaShortInterval) {
          triggerWatchViaShortInterval = false;
          node.log(
            `attempting reconnect to kubernetes watch (${
              kc.getCurrentCluster().server
            }${endpoint}) due to short interval trigger`
          );
          node.startWatch();
        }
      }, shortIntervalSeconds * 1000);
    } else {
      node.error("missing KubeConfig");
      node.status(statuses.misconfigured);
    }
  }
  RED.nodes.registerType("kubernetes-client-watch", KubernetesClientWatchNode);

  /**
   * TODO: support new option to 'continue' through the pages of responses
   *
   * @param {*} n
   */
  function KubernetesClientHttpNode(n) {
    RED.nodes.createNode(this, n);

    this.options = {};

    const node = this;
    node.lastMessageTimestamp = 0;
    this.kubernetesClientConfig = n.kubernetesClientConfig;
    this.kubernetesClientConfigNode = RED.nodes.getNode(
      this.kubernetesClientConfig
    );

    const kc = this.kubernetesClientConfigNode.kc;

    if (node.kubernetesClientConfig) {
      node.on("input", async function(msg, send, done) {
        node.status(statuses.sending);

        // support of 1.0+ and pre-1.0
        send =
          send ||
          function() {
            node.send.apply(node, arguments);
          };

        try {
          /**
           * Properties of the response include:
           *
           * statusCode
           * body
           * headers
           * request
           */
          let res = await kc.makeHttpRestRequest(msg);
          msg.payload = res.body;

          /**
           * try to add selfLink to involvedObject
           */
          if (
            ["Event", "EventList"].includes(msg.payload.kind) &&
            msg.payload.apiVersion == "v1"
          ) {
            try {
              switch (msg.payload.kind) {
                case "Event":
                  await kc.dressEventResource(msg.payload);

                  break;
                case "EventList":
                  await Promise.all(
                    msg.payload.items.map(async element => {
                      try {
                        return kc.dressEventResource(element);
                      } catch (err) {}
                    })
                  );

                  break;
              }
            } catch (err) {}
          }

          msg.kube = {};
          msg.kube.response = JSON.parse(JSON.stringify(res));
          msg.kube.config = {};
          msg.kube.config.cluster = kc.getCurrentCluster();
          msg.kube.config.context = kc.getCurrentContext();
          msg.kube.config.user = kc.getCurrentUser();
          msg.kube.client = kc;
          send(msg);
          node.status(statuses.blank);
          if (done) {
            done();
          }
        } catch (err) {
          const status = JSON.parse(JSON.stringify(statuses.error));
          status.text = status.text + ": " + err;
          node.status(status);

          // Report back the error
          if (done) {
            // Use done if defined (1.0+)
            done(err);
          } else {
            // Fallback to node.error (pre-1.0)
            node.error(err, msg);
          }
        }
      });
    } else {
      node.error("missing KubeConfig");
      node.status(statuses.misconfigured);
    }
  }
  RED.nodes.registerType("kubernetes-client-http", KubernetesClientHttpNode);
};
